man/tss2_createnv.1.md
% tss2_createnv(1) tpm2-tools | General Commands Manual
%
% APRIL 2019

# NAME

**tss2_createnv**(1) -

# SYNOPSIS

**tss2_createnv** [*OPTIONS*]

[common fapi references](common/tss2-fapi-references.md)

# DESCRIPTION

**tss2_createnv**(1) - This command creates an NV index in the TPM.

# OPTIONS

These are the available options:

  * **-p**, **\--path**=_STRING_:

    Path of the new NV space.

    The path is composed of three elements, separated by "/". An nvPath starts
    with "/nv". The second path element identifies the NV handle range
    for the nv object. This includes the following values:
    Owner, TPM, Platform, Endorsement_Certificate, Platform_Certificate,
    Component_OEM, TPM_OEM, Platform_OEM, PC-Client, Server,
    Virtualized_Platform, MPWG, Embedded. The third path element identifies
    the actual NV-Index using a meaningful name.

  * **-t**, **\--type**=_STRING_:

    Identifies the intended usage. Optional parameter.
    Types may be any comma-separated combination of:

        - "noda": Sets the noda attribute of a key or NV index.
        - "bitfield": Sets the NV type to bitfield.
        - "counter": Sets the NV type to counter.
        - "pcr": Sets the NV type to pcr-like behavior.
        - Hint: If none of the previous three keywords is provided a regular NV
          index is created.


  * **-s**, **\--size**=_INTEGER_:

    The size in bytes of the NV index to be created. Can be omitted if size can
    be inferred from the type; e.g. an NV index of type counter has a size of 8
    bytes.

  * **-P**, **\--policyPath**=_STRING_:

    Identifies the policy to be associated with the new NV space. Optional parameter.
    If omitted then no policy will be associated with the key.

    A policyPath is composed of two elements, separated by "/". A policyPath
    starts with "/policy". The second path element identifies the policy
    or policy template using a meaningful name.

  * **-a**, **\--authValue**=_STRING_:

    The new UTF-8 password. Optional parameter. If it is neglected then the user
    is queried interactively for a password. To set no password, this option
    should be used with the empty string (""). The maximum password size is
    determined by the digest size of the chosen name hash algorithm in the
    cryptographic profile (cf., **fapi-profile(5)**). For example, choosing
    SHA256 as hash algorithm, allows passwords of a maximum size of 32 characters.

[common tss2 options](common/tss2-options.md)

# EXAMPLE
```
tss2_createnv --authValue=abc --path=/nv/Owner/myNV --size=20 --type="noDa"
```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/fapi/tss2_createnv.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tools/fapi/tss2_template.h"

/* needed to conditionally free variable authValue */
static bool has_asked_for_password = false;

/* Context struct used to store passed commandline parameters */
static struct cxt {
    char     const *nvPath;
    char     const *nvTemplate;
    char           *authValue;
    uint32_t        size;
    char     const *policyPath;
} ctx;

/* Parse commandline parameters */
static bool on_option(char key, char *value) {
    switch (key) {
    case 'a':
        ctx.authValue = value;
        break;
    case 'P':
        ctx.policyPath = value;
        break;
    case 'p':
        ctx.nvPath = value;
        break;
    case 's':
        if (!tpm2_util_string_to_uint32 (value, &ctx.size)) {
            fprintf (stderr, "%s cannot be converted to an integer or is" \
                " larger than 2**32 - 1\n", value);
            return false;
        }
        break;
    case 't':
        ctx.nvTemplate = value;
        break;
    }
    return true;
}

/* Define possible commandline parameters */
static bool tss2_tool_onstart(tpm2_options **opts) {
    struct option topts[] = {
        {"path",       required_argument, NULL, 'p'},
        {"type",       required_argument, NULL, 't'},
        {"size",       required_argument, NULL, 's'},
        {"policyPath", required_argument, NULL, 'P'},
        {"authValue",  required_argument, NULL, 'a'},
    };
    return (*opts = tpm2_options_new ("P:a:p:s:t:", ARRAY_LEN(topts), topts,
                                      on_option, NULL, 0)) != NULL;
}

/* Execute specific tool */
static int tss2_tool_onrun (FAPI_CONTEXT *fctx) {
    /* Check availability of required parameters */
    if (!ctx.nvPath) {
        fprintf (stderr, "No NV path provided, use --path\n");
        return -1;
    }

    uint32_t size = 0;
    if (!ctx.size) {
        /* ctx.size is allowed to be zero if type is bitfield, pcr or
         * counter
         */
        if (!ctx.nvTemplate || !(strstr(ctx.nvTemplate, "bitfield") ||
            strstr(ctx.nvTemplate, "pcr") || strstr(ctx.nvTemplate, "counter"))) {
            fprintf (stderr, "Error: Either provide a type of \"bitfield\", "\
                "pcr\" or \"counter\" with --type or provide a size > 0 with "\
                "--size.\n");
            return -1;
        }
    }
    else {
        size = ctx.size;
    }

    /* If no authValue was given, prompt the user interactively */
    if (!ctx.authValue) {
        ctx.authValue = ask_for_password ();
        has_asked_for_password = true;
        if (!ctx.authValue){
            return 1; /* User entered two different passwords */
        }
    }

    /* Execute FAPI command with passed arguments */
    TSS2_RC r = Fapi_CreateNv(fctx, ctx.nvPath, ctx.nvTemplate,
        size, ctx.policyPath, ctx.authValue);
    if (r != TSS2_RC_SUCCESS){
        if(has_asked_for_password){
            free (ctx.authValue);
        }
        LOG_PERR ("Fapi_CreateNv", r);
        return 1;
    }

    if(has_asked_for_password){
        free (ctx.authValue);
    }

    return r;
}

TSS2_TOOL_REGISTER("createenv", tss2_tool_onstart, tss2_tool_onrun, NULL)
